### PR TITLE
Implement new Loop and Scan operators

### PR DESCRIPTION
**Pull request by ricardoV94**
Related to #189 

This PR implements a new low level `Loop` `Op` which can be easily transpiled to `Numba` (the Python perform method takes 9 lines, yay to not having to support `C` in the future).

It also implements a new higher level `Scan` `Op` which returns as outputs the last states + intermediate states of a looping operation. This `Op` cannot be directly evaluated, and must be rewritten as a `Loop` `Op` in Python/Numba backends. For the `JAX` backend it's probably fine to transpile directly from this representation into a `lax.scan` as the signatures are pretty much identical. That was not done in this PR.

The reason for the two types of outputs, is that they are useful in different contexts. Final states are sometimes all one needs, whereas intermediate states are generally needed for back propagation (not implemented yet). This allows us to choose which one (or both) of the outputs we want during compilation, without having to do complicated graph analysis. 

The existing `save_mem_new_scan` is used to convert a general scan into a `loop` that only returns the last computed state. It's... pretty complicated (although it also covers cases where more than 1 but less than all steps being requested, but OTOH it can't handle while loops #178):

https://github.com/pymc-devs/pytensor/blob/8ad33179b12a4c0207b2a654badc608e211e8bb9/pytensor/scan/rewriting.py#L1119

Taking that as a reference I would say the new conversion rewrite from Scan to Loop is much much simpler. Most of it is boilerplate code for defining the right trace inputs and new FunctionGraph

***

Both `Ops` expect a `FunctionGraph` as input. This should probably be created by a user-facing helper that accepts a callable like scan does now. ~~That was not done yet, as I first wanted to discuss the general design.~~ Done

## Design issues
~1. The current implementation of Loop assumes there are as many states as outputs of the inner function. This does not make sense for mapping or "filling" operations such as filling a tensor with random values. In one of the tests I had to create a dummy `x` input to accommodate this restriction. Should we use `NoneConst` to represent outputs that don't feed into the next state? I think there is something similar being done with the old `Scan` where the `outputs_info` must explicitly be `None` in these cases.~

2. Scan and Loop can now take random types as inputs (scan can't return it as a sequence). This makes random seeding much more explicit compared to the old Scan, which was based on default updates of shared variables. However it highlights the awkwardness of the random API when we want to access the next random state. Should we perhaps add a `return_rng_update` to `__call__`, so that it doesn't hide the next rng state output?

3. Do we want to be able to represent empty Loop / Sequences? If so, how should we go about that? `IfElse` is one option, but perhaps it would be nice to represent it in the same `Loop` `Op`?

4. What do we want to do in terms of inplacing optimizations?

## TODO
If people are on board with the approach
- [ ] Implement Numba dispatch
- [x] Implement JAX dispatch
- [ ] Implement L_op and R_op
- [x] Implement friendly user facing functions
- [ ] Decide on which meta-parameters to preserve (`mode`, `truncate_gradient`, `reverse` and so on)
- [x] Add rewrite that replaces `trace[-1]` by the first set of outputs (final state). That way we can keep the old API, while retaining the benefit of doing while Scans without tracing when it's not needed.

### CODE REVIEW

File: pytensor/loop/op.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+from pytensor.tensor.type_other import NoneTypeT
+
+
+def validate_loop_update_types(update):
+    assert update.outputs[0].type.dtype == "bool"
+    for input_state, output_state in zip(update.inputs, update.outputs[1:]):
+        assert input_state.type == output_state.type
+
+
+class Loop(Op):
```
Comment (ricardoV94) [inline pytensor/loop/op.py:21]: TODO: Add mixit `HasInnerGraph` so that we can see the inner graph in debug_print

File: pytensor/loop/basic.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    if non_sequences is None:
+        non_sequences = []
+    else:
+        if not isinstance(non_sequences, (tuple, list)):
+            non_sequences = [non_sequences]
+        non_sequences = [as_symbolic(n) for n in non_sequences]
+
+    # Note: Old scan order is sequences + init + non_sequences
+    inner_sequences = [s[0] for s in sequences]
+    inner_inputs = [i.type() for i in init_states + inner_sequences + non_sequences]
```
Comment (ricardoV94) [inline pytensor/loop/basic.py:46]: Hmm, I can't do this or else it won't be possible to reference the outer graph (e.g., to request gradients wrt non-sequences) as is don't by the jacobian function

Comment (aseyboldt) [reply pytensor/loop/basic.py:46]: I don't get it yet. Can you give an example of what you'd want to do, but can't?

**Comment by aseyboldt** [comment]
> The current implementation of Loop assumes there are as many states as outputs of the inner function. This does not make sense for mapping or "filling" operations such as filling a tensor with random values. In one of the tests I had to create a dummy x input to accommodate this restriction. Should we use NoneConst to represent outputs that don't feed into the next state? I think there is something similar being done with the old Scan where the outputs_info must explicitly be None in these cases.

Wouldn't a fill loop look something like this?

```
state = (pt.scalar(0), pt.empty(shape, dtype), rng)
def update(idx, values, rng):
    value, rng = rng.normal()  # not exactly the api...
    values = pt.set_subtensor(values[idx], value)
    return (idx + 1, values, rng, idx < maxval)
```

(and very much need inplace rewrites for good performance...)

> Scan and Loop can now take random types as inputs (scan can't return it as a sequence). This makes random seeding much more explicit compared to the old Scan, which was based on default updates of shared variables. However it highlights the awkwardness of the random API when we want to access the next random state. Should we perhaps add a return_rng_update to __call__, so that it doesn't hide the next rng state output?

Good question...
Don't know either :-)

> Do we want to be able to represent empty Loop / Sequences? If so, how should we go about that? IfElse is one option, but perhaps it would be nice to represent it in the same Loop Op?

I think one rewrite that get's easier with the if-else-do-while approach would be moving out loop independent computations. Let's say we have a loop like

```
x = bigarray...
if not_empty:
    val = 0
    do:
        val = (val + x.sum()) ** 2
    while val < 10

# rewrite to
x = bigarray...
if not_empty:
    val = 0
    x_sum = x.sum()
    do:
        val = (val + x_sum) ** 2
    while val < 10
```

we could move `x.sum()` out of the loop. But with a while loop we can't as easily, because we only want to do x.sum() if the loop is not empty, and where would we then put that computation?


> What do we want to do in terms of inplacing optimizations?

Well, I guess we really need those :-)
I'm thinking it might be worth it to copy the initial state, and then donate the state to the inner function? And I guess we need to make sure rewrites are actually running on inner graphs as well...

**Comment by ricardoV94** [comment]
> we could move x.sum() out of the loop. But with a while loop we can't as easily, because we only want to do x.sum() if the loop is not empty, and where would we then put that computation?

Why can't we move it even if it's empty? Sum works fine. Are you worried about Ops that we know will fail with empty inputs?

***

About the filling Ops, yeah I don't see it as a problem anymore. Just felt awkward to create the dummy input when translating from scan to loop. I am okay with it now

File: pytensor/loop/basic.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/loop/basic.py:46]: This: https://pytensor.readthedocs.io/en/latest/tutorial/gradients.html#computing-the-jacobian

Note the box below where they say it's not possible to take the gradient with respect to the sequences, so it must be passed as a non-sequence. It's the same kind of problem, except in my first approach here I made it impossible to do so with the non-sequences as well. My approach was to replace all inputs by dummy inputs before calling the user provided function. That means a macro like `grad` cannot see the original graph and take the right gradient.

This however means I have to allow users to reference other variables (e.g., the inputs to that graph expression), which they are unlikely to pass as non-sequences (how would they know?)

I will change that. Maybe I can change it so that even the sequences are not dummy so that the approach that didn't use to work with the old scan would also work fine here. Once we have the "whole" graph we should be able to truncate it correctly (i.e, pass the minimum subset of `non-sequences` with the helper that @ferrine wrote for `replace`?

This is messy but I don't see another way around it. A large part of the old Scan initialization is indeed trying to deal with this:
https://github.com/pymc-devs/pytensor/blob/4da56ecf6c08cd5c5b960116b101ba6dde84a0db/pytensor/scan/basic.py#L890-L905

Comment (ricardoV94) [reply pytensor/loop/basic.py:46]: Other discussion I found relevant for this:
https://groups.google.com/g/theano-users/c/fh9uYZnFj40/m/YpBQE5vtHQAJ

https://groups.google.com/g/theano-dev/c/h6G0QptsQOs/m/Hpx5B9EWO04J

Comment (aseyboldt) [reply pytensor/loop/basic.py:46]: The whole point of the reverse loop was that we don't have to worry about that kind of thing. Shouldn't you just be able to use the gradients from the loop?
I guess that would only work if Scan is a subclass of OpFromGraph though, instead of an independent op that only knows about its Loop representation after a rewrite?

**Comment by aseyboldt** [comment]
That would change the behavior. If we move it out and don't prevent it from being executed, things could fail for instance if there's an assert somewhere, or some other error happens during it's evaluation. Also, it could be potentially very costly (let's say "solve an ode").

(somehow I accidentally edited your comment instead of writing a new one, no clue how, but fixed now)

File: pytensor/loop/basic.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/loop/basic.py]: This is not about the gradients of the Scan operations, it's about using a Scan to take gradients of other variables (e.g., to fill in the jacobian matrix). Anyway I pushed a new commit that fixes it

**Comment by ricardoV94** [comment]
In my last commit, sequences are demoted from special citizens to just another constant input. The user facing helper creates the right graph with indexing that is passed to the user provided function.

I also don't convert the constant inputs to dummies before calling the user function, which allows the example in the documentation to work, including the one that didn't work before (because both are now equivalent under the hood :))

https://pytensor.readthedocs.io/en/latest/tutorial/gradients.html#computing-the-jacobian

I still need to mask the state variables as it doesn't make sense for the user function to introspect the graph before the initial state (as that one changes over time)

File: pytensor/loop/basic.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        n_steps, idx, *prev_states, *sequences, *non_sequences, *extra_fgraph_inputs
+    )
+    assert isinstance(scan_outs, list)
+    last_states = scan_outs[: scan_op.n_states]
+    traces = scan_outs[scan_op.n_states :]
+    # Don't return the inner index state
+    return last_states[1:], traces[1:]
+
+
+def map(
```
Comment (ferrine) [inline pytensor/loop/basic.py:134]: What about subclassing Scan into 

* `Map(Scan)`
* `Reduce(Scan)`
* `Filter(Scan)`

It will be easier to dispatch into optimized implementations

Comment (ricardoV94) [reply pytensor/loop/basic.py:134]: We can do that later, not convinced we need that yet

File: pytensor/loop/basic.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    # Replace None init by dummy empty tensors
+    prev_states = []
+    prev_inner_states = []
+    for i, (init_state, inner_state, next_state) in enumerate(
+        zip(init_states, inner_states, next_states)
+    ):
+        if init_state is None:
+            # next_state may reference idx. We replace that by the initial value,
+            # so that the shape of the dummy init state does not depend on it.
+            [next_state] = clone_replace(
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/loop/basic.py:94]: Why is that better?

**Comment by ricardoV94** [comment]
Added a simple JAX dispatcher, works in the few examples I tried

File: pytensor/link/jax/dispatch/loop.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        raise NotImplementedError(
+            "Scan ops with while condition cannot be transpiled JAX"
+        )
+
+    # Apply inner rewrites
+    # TODO: Not sure this is the right place to do this, should we have a rewrite that
+    #  explicitly triggers the optimization of the inner graphs of Scan?
+    update_fg = op.update_fg.clone()
+    rewriter = get_mode("JAX").optimizer
+    rewriter(update_fg)
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/loop.py:22]: This issues the annoying Supervisor Feature missing warning... gotta clean that up

File: pytensor/link/jax/dispatch/loop.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        def scan_fn(carry, _):
+            resume, *carry = jaxified_scan_inner_fn(*carry, *constants)
+            assert resume
+            carry = list(carry)
+            # Return states as both carry and output to be appended
+            return carry, [c for i, c in enumerate(carry) if i in traceable_states]
+
+        print(max_iters)
+        states, traces = jax.lax.scan(
+            scan_fn, init=list(states), xs=None, length=max_iters
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/loop.py:44]: Todo: Check we are not missing performance by not having explicit sequences.

Todo: When there are multiple sequences PyTensor defines n_steps as the shortest sequence. JAX should be able to handle this, but if not we could consider not allowing sequences/n_steps with different lengths in the Pytensor scan.

Then we could pass a single shape as n_steps after asserting they are the same?

**Comment by ricardoV94** [comment]
I just found out about [TypedLists](https://github.com/pymc-devs/pytensor/blob/491f93ed66f4b1a28b7b5920f749fdf684b61eaa/pytensor/typed_list/type.py) in PyTensor. That should allow us to trace any type of Variables, including RandomTypes :o: 

Pushed a couple of commits that rely on this.

**Comment by ricardoV94** [comment]
This Discourse thread is a great reminder of several Scan design issues that are fixed here: https://discourse.pymc.io/t/hitting-a-weird-error-to-do-with-rngs-in-scan-in-a-custom-function-inside-a-potential/13151/15
